### PR TITLE
TF-1068 Make a recent list for input of username

### PR DESCRIPTION
**Pull request by sherlockvn**
#1068 
- [x] save recent login and make a recent list for input of username

### Resolved

https://user-images.githubusercontent.com/43041967/197471563-7433eaf3-a9d1-4210-b323-44dd254735f2.mp4

https://user-images.githubusercontent.com/43041967/197509054-17d7bd8a-778d-4c2a-9cd2-83850ade5551.mp4

### CODE REVIEW

File: lib/features/caching/recent_login_username_cache_client.dart [hunk 1]
```diff
@@ -0,0 +1,126 @@
+import 'package:hive/hive.dart';
+import 'package:tmail_ui_user/features/caching/config/hive_cache_client.dart';
+import 'package:tmail_ui_user/features/login/data/model/recent_login_username_cache.dart';
+
+class RecentLoginUsernameCacheClient extends HiveCacheClient<RecentLoginUsernameCache> {
```
Comment (hoangdat) [inline lib/features/caching/recent_login_username_cache_client.dart:5]: please apply the new :)

**Comment by hoangdat** [comment]
1. should be focus to `password` field after select username

File: lib/features/login/presentation/base_login_view.dart [hunk 2]
```diff
@@ -106,17 +108,36 @@ abstract class BaseLoginView extends GetWidget<LoginController> {
...
-            ..textInputAction(TextInputAction.next)
-            ..keyboardType(TextInputType.emailAddress)
-            ..textDecoration((LoginInputDecorationBuilder()
-                ..setLabelText(AppLocalizations.of(context).email)
-                ..setHintText(AppLocalizations.of(context).email))
-              .build()))
-          .build()));
+      padding: const EdgeInsets.only(bottom: 24, right: 24, left: 24),
+      child: TypeAheadFormField<RecentLoginUsername>(
+      key: const Key('login_username_input'),
```
Comment (dab246) [inline lib/features/login/presentation/base_login_view.dart:113]: add 1 indent for for the elements of `TypeAheadFormField`

File: lib/features/login/presentation/login_controller.dart [hunk 3]
```diff
@@ -170,6 +178,7 @@ class LoginController extends BaseController {
   }
 
   void handleLoginPressed() {
+    _saveRecentLoginUsername();
```
Comment (dab246) [inline lib/features/login/presentation/login_controller.dart:181]: We should only save the email when it is in valid. Correct email format `x@b.com`

File: lib/features/login/data/datasource_impl/login_username_datasource_impl.dart [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        await _recentLoginUsernameCacheClient.insertItem(recentLoginUsername.username,
+            recentLoginUsername.toRecentLoginUsernameCache());
+      }
+    }).catchError((error) {
+      throw error;
+    });
+  }
+
+  bool _filterRecentLoginUsernameCache(
+      RecentLoginUsernameCache recentLoginUsernameCache, String? pattern) {
```
Comment (dab246) [inline lib/features/login/data/datasource_impl/login_username_datasource_impl.dart:50]: each param 1 line

File: lib/features/cleanup/domain/usecases/cleanup_recent_login_username_interactor.dart [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import 'package:core/presentation/state/failure.dart';
+import 'package:core/presentation/state/success.dart';
+import 'package:dartz/dartz.dart';
+import 'package:tmail_ui_user/features/cleanup/domain/model/recent_login_username_cleanup_rule.dart';
+import 'package:tmail_ui_user/features/cleanup/domain/repository/cleanup_repository.dart';
+import 'package:tmail_ui_user/features/cleanup/domain/state/cleanup_recent_login_username_cache_state.dart';
+
+class CleanupRecentLoginUsernameCacheInteractor {
```
Comment (dab246) [inline lib/features/cleanup/domain/usecases/cleanup_recent_login_username_interactor.dart:8]: Where is it used?

**Review by dab246 — COMMENTED** [review]
Quick review

**Comment by hoangdat** [comment]
it seems the app fills `email` after come from `url screeen`? @sherlockvn

**Review by hoangdat — APPROVED** [review]

**Review by ManhNTX — APPROVED** [review]
LGTM

**Review by dab246 — APPROVED** [review]
So good

**Pull request merged**